File: ProyectoCivitasJava/src/civitas/Jugador.java
/**
 * Proyecto Civitas
 * Sergio Hervás Cobo
 * Javier Suárez Lloréns
 * Programación y Diseño Orientado a Objetos
 * 2º B3- Grado en Ingeniería Informática
 * E.T.S.I. Informática y de Telecomunicación
 * Universidad de Granada
 */
package civitas;

import GUI.Dado;
import java.util.ArrayList;

public class Jugador implements Comparable<Jugador> {
    
    //ATRIBUTOS DE CLASE
    protected static int CasasMax = 4;
    protected static int CasasPorHotel = 4;
    protected static float PasoPorSalida = 1000f;
    protected static float PrecioLibertad = 200f;
    protected static int HotelesMax = 4;
    private static float  SaldoInicial = 7500f;
    
    //ATRIBUTOS DE INSTANCIA
    private String nombre;
    private int numCasillaActual;
    protected boolean encarcelado;
    private boolean puedeComprar;
    private float saldo;
    
    //ATRIBUTOS DE REFERENCIA
    private ArrayList <TituloPropiedad> propiedades;
    private SorpresaSalirCarcel salvoconducto;
    
    
    Jugador(String nombre){
        this.nombre = nombre;
        puedeComprar = false;
        saldo = SaldoInicial;
        numCasillaActual = 0;
        encarcelado = false;
        propiedades = new ArrayList<>();   
        salvoconducto = null;
    }
    
    protected Jugador (Jugador otro){
        
        salvoconducto = otro.salvoconducto;
        nombre = otro.nombre;
        puedeComprar = otro.puedeComprar;
        saldo = otro.saldo;
        numCasillaActual = otro.numCasillaActual;
        encarcelado = otro.encarcelado;
        propiedades = otro.propiedades;
        
    }
    
    
    boolean cancelarHipoteca (int ip){
     boolean result = false;
     if (!isEncarcelado()){
         if (existeLaPropiedad(ip)){
             TituloPropiedad propiedad = propiedades.get(ip);
             float cantidad = propiedad.getImporteCancelarHipoteca();
             boolean puedoGastar = puedoGastar(cantidad);
             
             if(puedoGastar){
                 result = propiedad.cancelarHipoteca(this);
                 
                 if(result)
                     Diario.getInstance().ocurreEvento(nombre + " cancelar la hipoteca de la propiedad " + ip + "\n" + propiedades.get(ip)+ "\n");
                 
             }
        }
     }
     return result;
    }
    
    
    int cantidadCasasHoteles(){
        int numero = 0;
        for (int i = 0; i < propiedades.size(); i++)
            numero += propiedades.get(i).cantidadCasasHoteles();
        return numero;
    }
    
    
    @Override
    public int compareTo(Jugador otro){
        return Float.compare(saldo, otro.getSaldo());
    }
    
    
    
    
    boolean comprar(TituloPropiedad titulo){
        boolean result = false;
        
        if(!encarcelado){
            if (puedeComprar){
                float precio = titulo.getPrecioCompra();
                
                if (puedoGastar(precio)){
                    result = titulo.comprar(this);
                    if (result){               
                        propiedades.add(titulo);
                        Diario.getInstance().ocurreEvento(this + " compra la propiedad " + titulo.getNombre() + "\n" + titulo + "\n");               
                    }
                    puedeComprar = false;
                }
            }
        }
        return result;
    }
    
    
    boolean construirCasa (int ip){
        boolean result = false;
        boolean puedoEdificarCasa = false;
        if (!encarcelado){
            boolean existe = existeLaPropiedad(ip);
            if (existe){
                TituloPropiedad propiedad = propiedades.get(ip);
                puedoEdificarCasa = puedoEdificarCasa(propiedad);
                if(puedoEdificarCasa){
                    result = propiedad.construirCasa(this);
                    if(result){
                        Diario.getInstance().ocurreEvento(nombre + " construye casa en la propiedad " + ip + "\n" + propiedades.get(ip)+ "\n");
                    }
                } 
            }
        } 
        return result;
    }
    
    
    boolean construirHotel(int ip){
        boolean result = false;
        if (!encarcelado){
            if(existeLaPropiedad(ip)){
             TituloPropiedad propiedad = propiedades.get(ip);
             boolean puedoEdificarHotel = puedoEdificarHotel(propiedad);
             float precio = propiedad.getPrecioEdificar();
             
             if(puedoGastar(precio)){
                 if(propiedad.getNumHoteles()<getHotelesMax()){
                     if (propiedad.getNumCasas() >= getCasasPorHotel()){
                         puedoEdificarHotel = true;
                         
                     }
                 }
             }
             
             if(puedoEdificarHotel){
                 result = propiedad.construirHotel(this);
                 int casasPorHotel = getCasasPorHotel();
                 propiedad.derruirCasas(casasPorHotel, this);
             }
                
                
            }            
        }
        
        if (result) {
            Diario.getInstance().ocurreEvento(nombre + " construye hotel en la propiedad " + ip + "\n" + propiedades.get(ip)+ "\n");
        }
        return result;
    }
    
    
    protected boolean debeSerEncarcelado(){
        boolean debeserlo = false;
        if(!isEncarcelado()){
            if(!tieneSalvoconducto()){
                debeserlo = true;
            }
            else{
                perderSalvoconducto();
                Diario.getInstance().ocurreEvento(nombre + " se libra de la cárcel");
            }
        }
        return debeserlo;
    }
    
    boolean enBancarrota(){
       boolean comprobacion = false;
       if(saldo <= 0)
           comprobacion = true;
       return comprobacion;
        
    }
    
    boolean encarcelar(int numCasillaCarcel){
        if(debeSerEncarcelado()){
            moverACasilla(numCasillaCarcel);
            encarcelado = true;
            Diario.getInstance().ocurreEvento(nombre + " ha sido encarcelado.");

        }
        return encarcelado;
        
    }
    
    private boolean existeLaPropiedad(int ip){
        return (ip>=0&&ip<propiedades.size());
    }
    
    
    //visibilidad cambiada a protected
    protected int getCasasMax(){
        return CasasMax;
    }
    
    
    //visibilidad cambiada a protected
    protected int getHotelesMax(){
        return HotelesMax;
    }
    
    int getCasasPorHotel(){
        return CasasPorHotel;
    }
    
    //cambiada visibilidad a publico
    public String getNombre(){ 
        return nombre;
    }
    
    int getNumCasillaActual(){
        return numCasillaActual;
    }
    
    private float getPrecioLibertad(){
        return PrecioLibertad;
    }
    
    private float getPremioPasoSalida(){
        return PasoPorSalida;
    }

    
    public ArrayList<TituloPropiedad> getPropiedades(){ //Debe ser publico para el método gestionar de VistaTextual
        return propiedades;
        
    }
    
    boolean getPuedeComprar(){
        return puedeComprar;        
    }
    
        //cambiada visibilidad a publico
    public float getSaldo(){
        return saldo;
    }
    
    
    boolean hipotecar(int ip){
        boolean result = false;
        
        if(!encarcelado){
            if (existeLaPropiedad(ip)){
                TituloPropiedad propiedad = propiedades.get(ip);
                result = propiedad.hipotecar(this);
                
                
            }
            
            if (result){
                Diario.getInstance().ocurreEvento(nombre + " hipoteca la propiedad " + ip + "\n" + propiedades.get(ip)+ "\n");
            }
        }
        return result;
    }
    
    public boolean isEncarcelado(){
        return encarcelado;
    }
    
    boolean modificarSaldo(float cantidad){
        saldo = saldo+cantidad;
        String evento;
        
        if (cantidad < 0){
            float devuelve = cantidad*(-2) + cantidad;
            evento = this + " pierde " + devuelve + ". ";
        } else {
            evento = this + " gana " + cantidad + ". ";
        }
        Diario.getInstance().ocurreEvento(evento);
        return true;
    }
    
    boolean moverACasilla(int numCasilla){
        boolean comprobacion = false;
        if(!isEncarcelado()){
            numCasillaActual = numCasilla;
            puedeComprar = false;
            Diario.getInstance().ocurreEvento("Te mueves a la casilla: " + numCasilla);
            comprobacion = true;
        }
        return comprobacion;
    }
    
    boolean obtenerSalvoconducto(SorpresaSalirCarcel sorpresa){
        boolean comprobacion = false;
        if(!isEncarcelado()){
            salvoconducto = sorpresa;
            comprobacion = true;
        }
        return comprobacion;
        
    }
    
    boolean paga(float cantidad){
        return modificarSaldo(cantidad*-1);
        }
    
    boolean pagaAlquiler(float cantidad){
        boolean comprobacion = false;
        if(!isEncarcelado()){
            comprobacion = paga(cantidad);
        }
        return comprobacion;        
    }
    
    boolean pagaImpuesto(float cantidad){
        boolean comprobacion = false;
        if(!isEncarcelado()){
            comprobacion = paga(cantidad);
        }
        
        return comprobacion;
    }
    
    boolean pasaPorSalida(){
        modificarSaldo(getPremioPasoSalida());
        Diario.getInstance().ocurreEvento("El jugador ha pasado por la salida.");
        return true;
    }
    
   private void perderSalvoconducto(){
        salvoconducto.usada();
        salvoconducto = null;     
    }
    
    boolean puedeComprarCasilla(){
       puedeComprar = true;
        if(isEncarcelado()){
           puedeComprar = false;
       }
       return puedeComprar;
    }
    
    private boolean puedeSalirCarcelPagando(){
        boolean comprobacion = false;
        if(saldo >= getPrecioLibertad()){
            comprobacion = true;
        }
        return comprobacion;
    }
    
    
    //visibilidad cambiada a protected
    protected boolean puedoEdificarCasa(TituloPropiedad propiedad){
        boolean puedoedificarcasa = false;
        float precio = propiedad.getPrecioEdificar();
        
        if (puedoGastar(precio) && propiedad.getNumCasas() < getCasasMax()){
            puedoedificarcasa = true;
        }
        return puedoedificarcasa;
    }
    
    
    //visibilidad cambiada a protected
    protected boolean puedoEdificarHotel(TituloPropiedad propiedad){
        boolean puedoedificarhotel = false;
        float precio = propiedad.getPrecioEdificar();
        if ( puedoGastar(precio) && propiedad.getNumHoteles() < getHotelesMax() && propiedad.getNumCasas() >= getCasasPorHotel()){
            puedoedificarhotel = true;
        }
        return puedoedificarhotel;
    }        
    
    //visibilidad cambiada a protected
    protected boolean puedoGastar(float precio){
        boolean comprobacion = false;
        if(!isEncarcelado()){
            comprobacion = saldo >= precio;
            
        }
        return comprobacion;
    }
    
    boolean recibe(float cantidad){
        boolean comprobacion = false;
        if(!isEncarcelado()){
            comprobacion = modificarSaldo(cantidad);
        }
        return comprobacion;
    }
    
    
    boolean salirCarcelPagando(){
        boolean comprobacion = false;
        if(isEncarcelado() && puedeSalirCarcelPagando()){
            paga(getPrecioLibertad());
            encarcelado = false;
            Diario.getInstance().ocurreEvento("El jugador ha salido de la carcel pagando");
            comprobacion = true;
        }
        return comprobacion;        
    }
    
    boolean salirCarcelTirando(){
        boolean comprobacion = false;
        if(Dado.getInstance().salgoDeLaCarcel()){
            encarcelado = false;
            Diario.getInstance().ocurreEvento("El jugador ha salido de la carcel tirando");
            comprobacion = true;
        }
        return comprobacion;
    }
    
    boolean tieneAlgoQueGestionar(){
        return propiedades.size()>0;
    }
    
    boolean tieneSalvoconducto(){
        return salvoconducto != null;  
    }
    
    @Override
    public String toString(){
      String salida = nombre + " ( " + saldo + " )" + " en la casilla " + numCasillaActual;
      
      return salida;
    } 
    
    boolean vender (int ip){
        boolean comprobacion = false;
        if(!isEncarcelado()){
            if(existeLaPropiedad(ip)){
                if(propiedades.get(ip).vender(this)){
                    Diario.getInstance().ocurreEvento("Se ha vendido la propiedad " + propiedades.get(ip).getNombre()+ "\n");
                    propiedades.remove(ip);
                    comprobacion = true;
                }
              
            }
     
        }
        return comprobacion;
    }
}


   

File: ProyectoCivitasRuby/lib/modelo/jugador.rb
#encoding: UTF-8
require_relative "diario"
require_relative "titulo_propiedad"
require_relative "dado"

module Civitas
  class Jugador
    
    #ATRIBUTOS DE INSTANCIA DE CLASE
    @casasmax = 4
    @hotelesmax = 4

    #ATRIBUTOS DE CLASE
    @@pasoporsalida = 1000.0
    @@preciolibertad = 200.0
    @@saldoinicial = 7000.0
    @@casasporhotel = 4
    
    
    #CONSTRUCTORES
    def initialize(nombre)
      #Atributos de instancia
      @nombre = nombre
      @propiedades = Array.new      #excepcion, atributo de referencia
      @salvoconducto = nil
      @encarcelado = false
      @num_casilla_actual = 0
      @puede_comprar = false 
      @saldo = @@saldoinicial
    end
    

    def copia(copia)
      @nombre = copia.nombre;
      @encarcelado = copia.encarcelado
      @num_casilla_actual = copia.num_casilla_actual
      @puede_comprar = copia.puede_comprar
      @saldo = copia.saldo
      @salvoconducto = copia.salvoconducto
      @propiedades = copia.propiedades
    end
    
    protected
    #LECTORES DE ATRIBUTOS DE CLASE
    attr_reader :pasoporsalida, :preciolibertad, :casasporhotel

    #LECTORES DE ATRIBUTOS DE INSTANCIA DE CLASE
    def self.casasmax
      @casasmax
    end
    
    def self.hotelesmax
      @hotelesmax
    end
    #MÉTODOS PÚBLICOS
    public 
    #LECTORES DE ATRIBUTOS DE INSTANCIA
    attr_reader :nombre, :num_casilla_actual, :saldo, :puede_comprar, :encarcelado, :propiedades, :salvoconducto
    


    
    def <=>(otro)
      @saldo <=> otro.saldo
    end
    
    
    def to_s
      salida = "#{@nombre} (#{@saldo}) en la casilla #{@num_casilla_actual}"
    end
   
    
    def isencarcelado()
      @encarcelado
    end
    
    
    def cancelar_hipoteca(ip)
      result = false
      if (@encarcelado == false)
        if(existe_la_propiedad(ip))
          propiedad = @propiedades.at(ip)
          cantidad = propiedad.get_importe_cancelar_hipoteca
          puedo_gastar = puedo_gastar(cantidad)
            
          if(puedo_gastar)
            result = propiedad.cancelar_hipoteca(self)
              
            if (result)
              Diario.instance.ocurre_evento("El jugador #{nombre} cancela la hipoteca de la propiedad #{ip}") 
            end 
          end   
        end      
      end
      result
    end
    
    
    def cantidad_casas_hoteles
      contador = 0
      for i in @propiedades
        contador+=i.cantidad_casas_hoteles();
      end
      contador
    end
    
    
    def comprar(titulo)
      result = false;
      if(@encarcelado == false)
        if(@puede_comprar)
          precio = titulo.preciocompra()
          if(puedo_gastar(precio))
            result = titulo.comprar(self)
            
            if(result)
              @propiedades.push(titulo)
              Diario.instance.ocurre_evento("El jugador #{nombre} compra la propiedad #{titulo.nombre}")
            end
            
            @puede_comprar = false;
            
          end
        end
      end
      result
    end
    
    
    def construir_casa(ip)
      result = false
      if(!@encarcelado)
        existe = existe_la_propiedad(ip)

        if(existe)
          propiedad = @propiedades.at(ip)
          puedo_edificar_casa =  puedo_edificar_casa(propiedad)
          
          if(puedo_edificar_casa)
            result = propiedad.construir_casa(self)
            
            if(result)
              Diario.instance.ocurre_evento("El jugador #{nombre} construye casa en la propiedad #{ip}")
            end
          end
          
        end
      end
      result
    end
    
    
    def construir_hotel(ip)
      result = false
      if (!@encarcelado)        
        if(existe_la_propiedad(ip))
          propiedad = @propiedades.at(ip)
          puedo_edificar_hotel = puedo_edificar_hotel(propiedad)

          
          if(puedo_edificar_hotel)
            result = propiedad.construir_hotel(self)
            propiedad.derruir_casas(@@casasporhotel, self)
            Diario.instance.ocurre_evento("El jugador #{nombre} construye hotel en la propiedad #{ip}")
            
          end
        end     
      end
      result
    end
    
    
    def en_bancarrota()
      comprobacion = false
      if(@saldo <= 0)
        comprobacion = true
      end
      
      comprobacion;
    end
    
    
    def encarcelar(num_casilla_carcel)
      if(debe_ser_encarcelado)
        mover_a_casilla(num_casilla_carcel)
        @encarcelado = true
        Diario.instance.ocurre_evento("El jugador ha sido encarcelado")   
      end
     
      @encarcelado
    end
    
    
    def hipotecar(ip)
      result = false
      if(!@encarcelado)
        if(existe_la_propiedad(ip))
          propiedad = propiedades.at(ip)
          result = propiedad.hipotecar(self)
          
        end
        
        if (result)
          Diario.instance.ocurre_evento("El jugador #{@nombre} hipoteca la propiedad #{ip}")
        end
      end
      result
    end
    
    
    def modificar_saldo(cantidad)
      @saldo = @saldo + cantidad;
      Diario.instance.ocurre_evento("Saldo modificado en #{cantidad}")
      true
    end
    
    
    def mover_a_casilla(numcasilla)
      comprobacion = false
      if (!@encarcelado)
        @num_casilla_actual = numcasilla
        @puede_comprar = false
        Diario.instance.ocurre_evento("El jugador #{nombre} se ha movido a la casilla #{numcasilla}")
        comprobacion = true
      end
    end
    
    
    def obtener_salvoconducto(sorpresa)
      comprobacion = false
      if(@encarcelado == false)
        @salvoconducto = sorpresa
        comprobacion = true
      end
    end
    
    
    def paga(cantidad)
      modificar_saldo(cantidad*-1)
    end
    
    
    def paga_alquiler(cantidad)
      comprobacion = false
      if(!@encarcelado)
        paga(cantidad)
      end
    end
    
    
    def paga_impuesto(cantidad)
      comprobacion = false
      if(!@encarcelado)
        paga(cantidad)
        comprobacion = true
      end
    end
    
    
    def pasa_por_salida
      modificar_saldo(@@pasoporsalida)
      Diario.instance.ocurre_evento("El jugador ha pasado por la salida")
      true
    end
    
    
    def puede_comprar_casilla
      comprobacion = false
      if(!@encarcelado)
        comprobacion = true
      end
      @puede_comprar = comprobacion
    end

    def recibe (cantidad)
      comprobacion = false
      if(!@encarcelado)
        modificar_saldo(cantidad)
      end
    end
    
    
    def salir_carcel_pagando
      comprobacion = false
      if(@encarcelado && puede_salir_carcel_pagando)
        paga(@@preciolibertad)
        @encarcelado = false
        Diario.instance.ocurre_evento("El jugador ha salido de la carcel pagando")
        comprobacion = true
      end
    end
    
    
    def salir_carcel_tirando
      comprobacion = Dado.instance.salgo_de_la_carcel
      if (comprobacion)
        @encarcelado = false
        Diario.instance.ocurre_evento("El jugador ha salido de la carcel tirando")
      end
      comprobacion
    end
    
    
    def tiene_algo_que_gestionar
      @propiedades.size>0
    end
    
    
    def tiene_salvoconducto
      return @salvoconducto!=nil
    end
    
    
    def vender(ip)
      comprobacion = false
      if (!@encarcelado)
        if (existe_la_propiedad(ip))
          vendida = @propiedades.at(ip).vender(self)
          if (vendida)
            Diario.instance.ocurre_evento("La propiedad #{@propiedades.at(ip).nombre} ha sido vendida")
            @propiedades.delete_at(ip)
            comprobacion = true
          end
        end
      end
      comprobacion
    end
    
    
    
    #MÉTODOS PROTEGIDOS
    protected  
    
    
    def debe_ser_encarcelado
      comprobacion = false
      if(!@encarcelado)
        if(!tiene_salvoconducto())
          comprobacion = true
        else
          perder_salvoconducto()
          Diario.instance.ocurre_evento("Salvoconducto usado. El usuario se libra de la cárcel")
        end
      end
      comprobacion
    end
  
    
    
    #MÉTODOS PRIVADOS
    private

    
    def existe_la_propiedad(ip)
      return (ip >= 0 && ip < @propiedades.size())
    end
    
    
    def perder_salvoconducto
      @salvoconducto.usada
      @salvoconducto = nil
    end
    
    
    def puede_salir_carcel_pagando
      return (@saldo >= @@preciolibertad)
    end
    
    
    def puedo_edificar_casa(propiedad)     
      puedo_edificar_casa = false
      precio = propiedad.precioedificar()
      
      if(puedo_gastar(precio) && propiedad.numcasas < Jugador.casasmax)
        puedo_edificar_casa = true
      end
          
      puedo_edificar_casa
    end
   
    
    def puedo_edificar_hotel(propiedad)
      puedo_edificar_hotel = false;
      precio = propiedad.precioedificar()
           
      if(puedo_gastar(precio) && propiedad.numhoteles < Jugador.hotelesmax && propiedad.numcasas >= @@casasporhotel)
        puedo_edificar_hotel = true
      end
      puedo_edificar_hotel
    end
    
    
    def puedo_gastar(precio)
      comprobacion = false
      if(!@encarcelado)
        comprobacion = @saldo>= precio
      end
    end
    

  end
end